app/coffee/ProjectManager.coffee

```diff
@@ -56,3 +57,35 @@ module.exports = ProjectManager =
 					callback new Error("Errors deleting docs. See log for details")
 				else
 					callback(null)
+
+	getProjectDocs: (project_id, projectStateHash, excludeVersions = {}, _callback = (error) ->) ->
+		timer = new Metrics.Timer("projectManager.getProjectDocs")
+		callback = (args...) ->
+			timer.done()
+			_callback(args...)
+
+		RedisManager.checkOrSetProjectState project_id, projectStateHash, (error, projectStateChanged) ->
+			return callback(error) if error?
+			# we can't return docs if project structure has changed
+			return callback Errors.ProjectStateChangedError("project state changed") if projectStateChanged
+			# project structure hasn't changed, return doc content from redis
+			RedisManager.getDocIdsInProject project_id, (error, doc_ids) ->
+				return callback(error) if error?
+				jobs = []
+				docs = []
+				for doc_id in doc_ids or []
+					do (doc_id) ->
+						jobs.push (cb) ->
+							# check the doc version first
+							RedisManager.getDocVersion doc_id, (error, version) ->
+								return cb(error) if error?
+								# skip getting the doc if we already have that version
+								return cb() if version is excludeVersions[doc_id]
+								# otherwise get the doc lines from redis
+								RedisManager.getDocLines doc_id, (error, lines) ->
+									return cb(error) if error?
+									docs.push {_id: doc_id, lines: lines, rev: version}
```

Comment:
We use `v` for the ShareJS version, and `rev` for the Mongo version (everything time it's flushed to Mongo). `rev` is used as the version numbers for keeping Dropbox etc too. The doc updater only cares about `v` though, so it should return this as `v`.